### PR TITLE
move code from `Oscar.jl/experimental/GModule/Misc.jl` here

### PR DESCRIPTION
**Pull request by ThomasBreuer**
The idea is to add this code to AbstractAlgebra.jl, and then to remove it from Oscar.jl.

Once we agree what shall be really added here, tests and documentation have to be added.

(I am quite sure that I have put certain parts into the wrong files, and that some functions have to be changed.)

### CODE REVIEW

**Comment by lgoettgens** [comment]
We need to adjust the signature here at least a little bit to not have something like https://github.com/oscar-system/Oscar.jl/issues/3576 again.

**Comment by thofma** [comment]
Yes, I wonder how we plan to do this? This is breaking for Oscar 1.0 and master. We cannot move methods around in a backwards compatible way.

**Comment by lgoettgens** [comment]
> Yes, I wonder how we plan to do this? This is breaking for Oscar 1.0 and master. We cannot move methods around in a backwards compatible way.

We can always put something like this into a "technically breaking" release of AA/Nemo/Hecke etc. as we plan to do with the printing changes anyway. Oscar 1.0 will never know about this, Oscar master just needs to get these functions removed in the same commit the AA compat is adapted.

**Comment by fingolfin** [comment]
We are accumulating a bunch of "technically breaking" changes right now, so let's just label this as `breaking`, too, and make a new breaking release as soon as we have the (super)compact bits here cleaned up, too?

**Comment by fieker** [comment]
On Wed, Apr 17, 2024 at 08:17:14AM -0700, Tommy Hofmann wrote:
> Yes, I wonder how we plan to do this? This is breaking for Oscar 1.0 and master. We cannot move methods around in a backwards compatible way.
> 
I'd say it depends on the definition. At the end of the move, nothing
has changed (to the user)...
OK, in AA, Nemo, Hecke there might be aditional functions, but in Oscar
all that was there before is there, just the location of the
implementation has changed

**Comment by ThomasBreuer** [comment]
The tests failed.
When I remove a "harmless" `matrix` method then the failures disappear in my local installation.
(I do not understand what is happening here.)

**Comment by fingolfin** [comment]
Just to explain why this is "technically" breaking: because this will cause warnings in OSCAR of the kind "warning, method redefined in file X line Y, was previously defined in AbstractAlgebra; this may break precompilation". And then break CI tests, *even though the methods we add here are identical to the ones in OSCAR*. But Julia does not know this and panics.

File: src/generic/DirectSum.jl [hunk 1]
```diff
@@ -20,6 +20,10 @@ base_ring(N::DirectSumModule{T}) where T <: RingElement = base_ring(N.m[1])
 
 base_ring(v::DirectSumModuleElem{T}) where T <: RingElement = base_ring(v.parent)
 
+is_free(M::DirectSumModule) = all(is_free, M.m)
```
Comment (fingolfin) [inline src/generic/DirectSum.jl:23]: Actually, isn't this mathematically wrong? That is: if all summands are free, then M is free. But M might be free even though its summands are not (keyword: projective modules, etc.)

File: src/generic/DirectSum.jl [hunk 2]
```diff
@@ -220,6 +224,11 @@ function direct_sum(m::Vector{<:AbstractAlgebra.FPModule{T}}) where T <: RingEle
    return M, inj, pro
 end
 
+function direct_sum(M::DirectSumModule{T}, N::DirectSumModule{T}, mp::Vector{ModuleHomomorphism{T}}) where T
```
Comment (lgoettgens) [inline src/generic/DirectSum.jl:227]: this function should have `hom` in its name. In Oscar, we call the similar function that lifts homs to a tensor product `hom_tensor`. For direct sums, there is the equivalent for Lie algebra modules as `hom_direct_sum`, but we can adapt the name of that in Oscar to match this one here. But IMO it should start with `hom_*`

**Comment by ThomasBreuer** [comment]
> what was the error you saw but did not understand? Perhaps if you can quote it here somebody has an idea?

When the `matrix(phi::IdentityMap{<:AbstractAlgebra.FPModule})` method in `src/generic/Map.jl` gets installed, I get the following error.

```
julia> using AbstractAlgebra
[ Info: Precompiling AbstractAlgebra [c3fe647b-3220-5bb0-a1ea-a7954cac585d]

julia> M = free_module(ZZ, 2)
Free module of rank 2 over integers

julia> a = [ZZ(1), ZZ(2)]
2-element Vector{BigInt}:
 1
 2

julia> M(a)
ERROR: MethodError: no method matching matrix(::AbstractAlgebra.Integers{BigInt}, ::Int64, ::Int64, ::Vector{BigInt})
Stacktrace:
 [1] (::AbstractAlgebra.Generic.FreeModule{BigInt})(a::Vector{BigInt})
   @ AbstractAlgebra.Generic /export/home/sam/julia/packages/AbstractAlgebra.jl/src/generic/FreeModule.jl:132
 [2] top-level scope
   @ REPL[4]:1

julia> @which M(a)
(M::AbstractAlgebra.Generic.FreeModule{T})(a::Vector{S}) where {T<:NCRingElement, S<:RingElement} in AbstractAlgebra.Generic at /export/home/sam/julia/packages/AbstractAlgebra.jl/src/generic/FreeModule.jl:129
```

**Comment by fieker** [comment]
On Thu, Apr 18, 2024 at 07:20:35AM -0700, Lars Göttgens wrote:
> @lgoettgens commented on this pull request.
> 
> 
> 
> > @@ -220,6 +224,11 @@ function direct_sum(m::Vector{<:AbstractAlgebra.FPModule{T}}) where T <: RingEle
>     return M, inj, pro
>  end
>  
> +function direct_sum(M::DirectSumModule{T}, N::DirectSumModule{T}, mp::Vector{ModuleHomomorphism{T}}) where T
> 
> this function should have `hom` in its name. In Oscar, we call the similar function that lifts homs to a tensor product `hom_tensor`. For direct sums, there is the equivalent for Lie algebra modules as `hom_direct_sum`, but we can adapt the name of that in Oscar to match this one here. But IMO it should start with `hom_*`
I am happy to discuss this, but in Oscar, originally it was even just
called sum. As with all this things, they started with abelian groups in
Hecke...
But I'd be happy to dicuss names. We have
 - a matrix of homs from Mi to Nj
 - a vector from Mi to Ni
 - a vector of homs from M to Ni
So many functions, so little names...

**Comment by fieker** [comment]
On Thu, Apr 18, 2024 at 04:30:37PM +0200, ***@***.*** wrote:
> On Thu, Apr 18, 2024 at 07:20:35AM -0700, Lars Göttgens wrote:
> > @lgoettgens commented on this pull request.
> > 
> > 
> > 
> > > @@ -220,6 +224,11 @@ function direct_sum(m::Vector{<:AbstractAlgebra.FPModule{T}}) where T <: RingEle
> >     return M, inj, pro
> >  end
> >  
> > +function direct_sum(M::DirectSumModule{T}, N::DirectSumModule{T}, mp::Vector{ModuleHomomorphism{T}}) where T
> > 
> > this function should have `hom` in its name. In Oscar, we call the similar function that lifts homs to a tensor product `hom_tensor`. For direct sums, there is the equivalent for Lie algebra modules as `hom_direct_sum`, but we can adapt the name of that in Oscar to match this one here. But IMO it should start with `hom_*`
> I am happy to discuss this, but in Oscar, originally it was even just
> called sum. As with all this things, they started with abelian groups in
Sorry, it was called hom

File: src/generic/Map.jl [hunk 3]
```diff
@@ -91,6 +91,8 @@ end
 
 Base.inv(f::AbstractAlgebra.Map(AbstractAlgebra.IdentityMap)) = f
 
+#matrix(phi::IdentityMap{<:AbstractAlgebra.FPModule}) = identity_matrix(base_ring(domain(phi)), dim(domain(phi)))
```
[the remark this answers is not in the document]
Comment (ThomasBreuer) [reply src/generic/Map.jl:94]: Thanks.
(I had expected a warning from Julia in this situation.)

File: src/generic/DirectSum.jl (hunk 1 above)
Comment (ThomasBreuer) [reply src/generic/DirectSum.jl:23]: @fieker In which situation was `is_free(::DirectSumModule)` missing?

**Comment by fieker** [comment]
On Fri, Apr 19, 2024 at 01:39:55AM -0700, Thomas Breuer wrote:
> @ThomasBreuer commented on this pull request.
> 
> 
> 
> > @@ -20,6 +20,10 @@ base_ring(N::DirectSumModule{T}) where T <: RingElement = base_ring(N.m[1])
>  
>  base_ring(v::DirectSumModuleElem{T}) where T <: RingElement = base_ring(v.parent)
>  
> +is_free(M::DirectSumModule) = all(is_free, M.m)
> 
> @fieker In which situation was `is_free(::DirectSumModule)` missing?
> 
Can't tell anymore. It might also now be supported by a generic is_free
for FPModule over any field
The only tricky case would be for modules over Z where due to torsion
freeness is not automatic

**Comment by thofma** [comment]
If it is only applied to finitely generated/presented modules over fields or PIDs, then the method is correct: $M \oplus N$ is free if and only if $M$ and $N$ are free.

File: src/Module.jl [hunk 4]
```diff
@@ -27,6 +29,23 @@ function check_parent(M::FPModuleElem{T}, N::FPModuleElem{T}) where T <: RingEle
    parent(M) !== parent(N) && error("Incompatible modules")
 end
 
+gen(M::FPModule, i::Int) = M[i]
```
[the remark this answers is not in the document]
Comment (ThomasBreuer) [reply src/Module.jl:32]: The type `AbstractAlgebra.FPModule` has exactly five subtypes.
```
julia> subtypes(AbstractAlgebra.FPModule)
5-element Vector{Any}:
 AbstractAlgebra.Generic.DirectSumModule
 AbstractAlgebra.Generic.FreeModule
 AbstractAlgebra.Generic.QuotientModule
 AbstractAlgebra.Generic.SNFModule
 AbstractAlgebra.Generic.Submodule
```
Each of them has its own `gen` method. Thus I do not see why this `gen` method was missing. I will remove it.

File: src/generic/DirectSum.jl (hunk 1 above)
Comment (fingolfin) [reply src/generic/DirectSum.jl:23]: @thofma points out:
> If it is only applied to finitely generated/presented modules over fields or PIDs, then the method is correct: $M \oplus N$ is free if and only if $M$ and $N$ are free.

This is of course true, but is there a way we can use this knowledge to improve the code above? I mean we could certainly restrict it to modules over fields; or over AA "integers", and AA univariates polynomials over fields, and perhaps  a few more. But we can't do this generally here (e.g. not for fields over `Nemo.ZZRing`). But maybe we don't, it depends on what the application here is?

Comment (thofma) [reply src/generic/DirectSum.jl:23]: No, but it is the same for all the functionality of things involving `FPModule`.

File: src/Module.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (fingolfin) [reply src/Module.jl:32]: Maybe there is a misunderstanding here -- there definitely should be a `gen(M::FPModule, i::Int)` method. But it should not be implemented as it is here, but rather it should replace the `getindex(::FPModule)` method which presumably is defined elsewhere in this file.

File: src/generic/DirectSum.jl (hunk 1 above)
Comment (fingolfin) [reply src/generic/DirectSum.jl:23]: Aha. The documentation indeed says:
```
# Free Modules and Vector Spaces

AbstractAlgebra allows the construction of free modules of any rank over any
Euclidean ring and the vector space of any dimension over a field. 
```
but that's the only place in `docs/src/free_module.md` mentioning Euclidean 
and it then proceeds to say
```
AbstractAlgebra provides generic types for free modules and vector spaces,
via the type `FreeModule{T}` for free modules, where `T`
is the type of the elements of the ring $R$ over which the module is built.
```
which to makes it not at all clear whether or not e.g. `FreeModule{T}` is only supposed to be used over Euclidean rings.

And `docs/src/direct_sum.md` does not mention Euclidean/principal/PID at all.


But `docs/src/module.md` then goes on to say this
```
The generic code provided by AbstractAlgebra will only work for modules over
euclidean domains.
```
which ought to settle it, except it *then* proceeds in the very next sentence to say
```
Free modules can be built over both commutative and noncommutative rings. Other
types of module are restricted to fields and euclidean rings.
```
which again leaves it completely unclear to me what is supposed to work in which case or not.

I think adding an `is_free` method that knowingly returns wrong results sometimes is problematic. At the very least its docstring then should contain a big warning?

Better would be if we could just restrict it -- e.g. if this only for direct sums over integers or fields it'd be fine. But that begs the question where is even used -- @fieker?


Alternatively I could also live with such a method:
```
function is_free(M::DirectSumModule)
  f = all(is_free, M.m)
  @req f || is_euclidean(base_ring(M)) "is_free for direct sums over non-euclidean domains not supported"
  return f
end

is_euclidean(::Field) = true
is_euclidean(::ZZRing) = true
is_euclidean(::PolyRing{<:Field}) = true
is_euclidean(::Ring) = error("cannot tell if this ring is Euclidean")
```

**Comment by fingolfin** [comment]
Maybe we can drop the `is_free` method from this PR for now / move it to a separate PR, so that we can proceed with merging the rest?

File: src/Module.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (lgoettgens) [reply src/Module.jl:32]: > Maybe there is a misunderstanding here -- there definitely should be a `gen(M::FPModule, i::Int)` method. 

No, every concrete subtype of `FPModule` needs to define one on its own

File: src/Module.jl [hunk 5]
```diff
@@ -27,6 +29,23 @@ function check_parent(M::FPModuleElem{T}, N::FPModuleElem{T}) where T <: RingEle
...
+
+is_finite(M::FPModule{<:FinFieldElem}) = true
+
+function is_sub_with_data(M::FPModule{T}, N::FPModule{T}) where T <: RingElement
```
Comment (lgoettgens) [inline src/Module.jl:36]: maybe rename to `is_sub_with_inclusion`?

File: src/Module.jl (hunk 4 above)
[the remark this answers is not in the document]
Comment (fingolfin) [reply src/Module.jl:32]: Ahhh, when I wrote this, I was under the impression that `FPModule` is a concrete type. Should have looked.

So yeah, it should just go.

File: src/Module.jl [hunk 6]
```diff
@@ -27,6 +29,23 @@ function check_parent(M::FPModuleElem{T}, N::FPModuleElem{T}) where T <: RingEle
    parent(M) !== parent(N) && error("Incompatible modules")
 end
 
+gen(M::FPModule, i::Int) = M[i]
+
```
Comment (fingolfin) [inline src/Module.jl:33]:
```suggestion
```

File: src/generic/DirectSum.jl [hunk 7]
```diff
@@ -20,6 +20,10 @@ base_ring(N::DirectSumModule{T}) where T <: RingElement = base_ring(N.m[1])
 
 base_ring(v::DirectSumModuleElem{T}) where T <: RingElement = base_ring(v.parent)
 
+is_free(M::DirectSumModule) = all(is_free, M.m)
+
```
Comment (fingolfin) [inline src/generic/DirectSum.jl:24]: Let's remove this here for now so we can get the rest merged.
```suggestion
```

**Comment by fingolfin** [comment]
I've remove the `gen(::FPModule, ::Int)` method and the problematic `is_free` method. If tests pass we can merge this.

**Pull request merged**